### PR TITLE
Session-based API

### PR DESCRIPTION
**Pull request by jerith**
This changes the signature of all functions that take a `token` and has them take a `session` instead. Like `token`, `session` is an abstract type.

For backwards compatibility, `token_of_string` now returns a session instead. In order to (optionally) provide a different `base_url`, you need to use the new `make_session` function instead.

The two API functions that didn't take a `token` (`api_test` and `oauth_access`) now accept an optional `?base_url` parameter and are otherwise unchanged.

The `slack_notify` command line tool now accepts a `--base-url` option (which does the obvious thing), mostly to demonstrate usage.

### CODE REVIEW

**Comment by jerith** [comment]
I've added basic support for caching identifiers (channel, user, group) in the session, but I didn't want to make this branch bigger (and harder to review) than it already is, so I put that in its own branch: https://github.com/jerith/slacko/tree/session-cache
Comparison with this branch: https://github.com/jerith/slacko/compare/session-based...jerith:session-cache

The only change to the public API is the new `flush_session_caches` function, but all identifier lookups now cache the complete mapping for that identifier type for one minute. It would be useful to update the cache automatically whenever `channels_list` etc. is called (and update or invalidate it in `channels_rename` etc.), but I haven't done that yet.

I'd prefer to keep the caching stuff out of this PR in the interests of breaking the work into more manageable chunks, but I'll continue working on it in parallel and rebasing as this PR changes.

File: src/slacko.ml [hunk 1]
```diff
@@ -834,7 +841,11 @@ let string_of_presence = function
   | Away -> "away"
 
 (* Slacko API helper methods *)
-let token_of_string = identity
+let make_session ?(base_url=default_base_url) token = {
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: How about `start_session`? This makes more sense in a Session context, but maybe `make_` makes more sense in an OCaml context (with `Make` functors). Or maybe `create_session`?

File: src/slacko.ml [hunk 2]
```diff
@@ -584,19 +590,23 @@ type history_result = [
...
 (* internal *)
+let default_base_url = "https://slack.com/api/"
+
 let optionally_add key value uri = match value with
   | None -> uri
   | Some value -> Uri.add_query_param' uri (key, value)
 
 let definitely_add key value = optionally_add key (Some value)
 
+let unauthed_endpoint ?(base_url=default_base_url) e =
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: I'd give `e` a name like `method` since that's what Slack calls it in the documentation.

File: src/slacko.ml [hunk 3]
```diff
@@ -584,19 +590,23 @@ type history_result = [
...
+  base_url ^ e
+  |> Uri.of_string
+
+let endpoint session e =
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: This would be nicer with the arguments flipped, that way the session can be threaded with `|>` and I'd argue that it's more useful to curry over the `e` (method) than over the session.

File: src/slacko.ml [hunk 4]
```diff
@@ -1593,14 +1562,13 @@ let search base token ?sort ?sort_dir ?highlight ?count ?page query_ =
     | #bot_error as res -> res
     | _ -> `Unknown_error
 
-let search_all = search @@ endpoint "search.all"
-let search_files = search @@ endpoint "search.files"
-let search_messages = search @@ endpoint "search.messages"
+let search_all session = search session @@ endpoint session "search.all"
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Could be done nicer with supplying the curried `endpoint "search.all"` and then calling `base session` inside, which removes some duplication.

File: src/slacko.mli [hunk 5]
```diff
@@ -631,8 +633,11 @@ type history_result = [
 (** To build the types required in the API calls, you can use these helper
     functions. *)
 
-(** Converts a string into a token. *)
-val token_of_string: string -> token
+(** Create a session from a token string and an optional base_url. *)
+val make_session: ?base_url:string -> string -> session
+
+(** Deprecated wrapper for backcompat. *)
+val token_of_string: string -> session
```
Comment (Leonidas-from-XIV) [inline src/slacko.mli]: Mark this `[@@ocaml.deprecated]`?

File: src/slacko.ml [hunk 6]
```diff
@@ -599,12 +599,16 @@ let optionally_add key value uri = match value with
 
 let definitely_add key value = optionally_add key (Some value)
 
-let unauthed_endpoint ?(base_url=default_base_url) e =
+let unauthed_endpoint ~base_url e =
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Why was this change necessary? It looks to me like it should've worked out of the box.

Comment (jerith) [inline src/slacko.ml]: Having the default set there made it possible to accidentally forget to pass it in one of the functions that doesn't take a `session` and thus requires an optional `base_url` arg. I found this while running some tests through a local mitmproxy (to more easily inspect the responses slack was returning) because I'd forgotten to pass `base_url` to `unauthed_endpoint` somewhere and the request wasn't showing up in my proxy.

File: src/slacko.ml (hunk 2 above)
Comment (jerith) [inline src/slacko.ml]: Will do.

File: src/slacko.ml (hunk 3 above)
Comment (jerith) [inline src/slacko.ml]: Good point. I pretty much just made `session` the first arg of anything that took it because that was sensible for most of the functions and I didn't really think about this case. :-)

File: src/slacko.ml (hunk 1 above)
Comment (jerith) [inline src/slacko.ml]: Any of those work, and I don't have particularly strong opinions on which. `start_session` seems slightly more sensible than the others, so I'll change to that.

File: src/slacko.mli (hunk 5 above)
Comment (jerith) [inline src/slacko.mli]: I didn't know that existed. Thanks. :-)

File: src/slacko.ml (hunk 4 above)
Comment (jerith) [inline src/slacko.ml]: Sounds reasonable. :-)

File: src/slacko.ml (hunk 2 above)
Comment (jerith) [inline src/slacko.ml]: Turns out `method` is a reserved word.

File: src/slacko.ml [hunk 7]
```diff
@@ -599,16 +599,16 @@ let optionally_add key value uri = match value with
 
 let definitely_add key value = optionally_add key (Some value)
 
-let unauthed_endpoint ~base_url e =
+let unauthed_endpoint ~base_url meth =
```
Comment (Leonidas-from-XIV) [inline src/slacko.ml]: Admittedly a bit pedantic, but I'd rather name it `method'` or `method_` since `meth` has kinda weird connocations.

Comment (jerith) [inline src/slacko.ml]: Works for me. I don't like trailing underscores particularly, but the apostrophe is fine.
